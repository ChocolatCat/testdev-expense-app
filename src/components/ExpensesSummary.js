import React from 'react';
import numeral from 'numeral';

const ExpensesSummary = ({ expensesCount = 0, expensesTotal =0 }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpenses = numeral(expensesTotal).format('$0,0[.]00');
    return (
        <div>
            <h1>Viewing {expensesCount} {expenseWord} totalling {formattedExpenses}.</h1>
        </div>
    );
};

export default ExpensesSummary;