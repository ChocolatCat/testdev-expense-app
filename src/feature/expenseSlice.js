import { createSlice } from '@reduxjs/toolkit'
import {nanoid} from 'nanoid';

//Redux toolkit syntax
//Name defines the name in the store, initial state initializes the value and reducers are the actions that can manipulate the value
//Actions are defined by types, using the name as the first thing then a / to define the action. ie. counter/increment
//Reducers are pure functions - output is only determined by the input
//They never change state or action

export const expenseSlice = createSlice({
    name: 'expenses',
    initialState: [],
    reducers: {
        addExpense: {
            reducer: (state, action) => {
                //destructuring to work easily with values
                const {id, createdAt, description, note, amount} = action.payload;
                const expense = {
                    id: id,
                    description: description,
                    note: note,
                    amount: amount,
                    createdAt: createdAt
                };
                state.push(expense);
            },
            //We prepare the elements to work on
            prepare: (expense) => {
                //We generate an ID
                const id = nanoid();
                //We grab the elements sent to the reducer. We also prepare default values
                const {description = "No description", note = "No note", amount = 0, createdAt = 0} = expense;
                return {payload: {id, createdAt, description, note, amount, createdAt}};
            }
        },
        removeExpense: (state, {payload}) => {
            //We filter the element with the corresponding ID
            return state.filter(({id}) => id !== payload);
        },
        editExpense: (state, {payload}) => {
            return state.map( (expense) => {
                if(expense.id === payload.id){
                    return {...expense, ...payload.updates};
                }
                else{
                    return expense;
                }
            });
        }
    }
});

export const {addExpense, removeExpense, editExpense, expensesTotal} = expenseSlice.actions;

export default expenseSlice.reducer;