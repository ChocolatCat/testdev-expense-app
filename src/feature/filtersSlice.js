import { createSlice } from '@reduxjs/toolkit'
import { startOfMonth, lastDayOfMonth } from 'date-fns';

//Redux toolkit syntax
//Name defines the name in the store, initial state initializes the value and reducers are the actions that can manipulate the value
//Actions are defined by types, using the name as the first thing then a / to define the action. ie. counter/increment
//Reducers are pure functions - output is only determined by the input
//They never change state or action

const filtersSlice = createSlice({
    name: 'filters',
    initialState: {
        text: '',
        sortBy: 'date',
        startDate: Date.parse(startOfMonth(Date.now())),
        endDate: Date.parse(lastDayOfMonth(Date.now()))
    },
    reducers: {
        setTextFilter: (state, action) => {
            state.text = action.payload ? action.payload : "";
        },
        sortByDate: (state) => {
            state.sortBy = 'date';
        },
        sortByAmount: (state) => {
            state.sortBy = 'amount';
        },
        setStartDate: (state, action) => {
            state.startDate = action.payload;
        },
        setEndDate: (state, action) => {
            state.endDate = action.payload;
        }
    }
});

export const {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} = filtersSlice.actions;

export default filtersSlice.reducer;