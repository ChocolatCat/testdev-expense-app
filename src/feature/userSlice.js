import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import {auth, googleAuthProvider} from '../firebase/firebase';
import { signInWithRedirect, signOut } from 'firebase/auth';

export const login = createAsyncThunk('user/login', 
async (arg, thunkAPI) => {
    signInWithRedirect(auth, googleAuthProvider).then((result) => {
        return true;
    });
});

export const logout = createAsyncThunk('user/logout', 
async (arg, thunkAPI) => {
    signOut(auth).then(() => {
        return false;
    });
});

const userSlice = createSlice({
    name: 'user',
    initialState: {
        isLogged: false,
        uid: undefined
    },
    reducers: {
        setLogStatus(state, {payload}) {
            state.isLogged = payload;
        },
        setUid(state, {payload}){
            state.uid = payload;
        }
    },
    extraReducers: (builder) => {
        builder.addCase(login.fulfilled, (state, {payload}) => {
            state.isLogged = payload;
        }),
        builder.addCase(login.rejected, (state, {payload}) => {
            state.isLogged = payload;
        });
    }
});

export const { setLogStatus, setUid } = userSlice.actions;

export default userSlice.reducer;