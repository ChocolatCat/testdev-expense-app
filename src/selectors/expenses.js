//import {createSelector} from '@reduxjs/toolkit';
import {isAfter, isBefore, isSameDay} from 'date-fns';

const selectExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = expense.createdAt;
        const startDateMatch = startDate ?  isAfter(createdAtMoment, startDate) || isSameDay(startDate, createdAtMoment) : true;
        const endDateMatch = endDate ? isBefore(createdAtMoment, endDate) || isSameDay(endDate, createdAtMoment) : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt > b.createdAt ? 1 : -1
        }
        if(sortBy === 'amount'){
            return a.amount > b.amount ? 1 : -1;
        }
    });  
};

const selectExpenseById = (expenses, id) => {
    return expenses.find( (expense) => {
        return expense.id === id;
    });
};

const selectExpensesTotal = (expenses) => {
    return expenses
        .map( (expense) => expense.amount )
        .reduce( (sum, value) => sum+value, 0);
}

export {selectExpenses, selectExpenseById, selectExpensesTotal};