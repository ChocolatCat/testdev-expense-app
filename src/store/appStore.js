import {configureStore} from '@reduxjs/toolkit';
import expenseReducer from '../feature/expenseSlice';
import filtersReducer from '../feature/filtersSlice';

export default () => {
    const store = configureStore({
        reducer:{
            expenses: expenseReducer,
            filters: filtersReducer
        }
    });
    return store;
}