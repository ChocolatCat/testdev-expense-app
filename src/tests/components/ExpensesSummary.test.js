/** @jest-environment jsdom */

import React from 'react';
import ExpensesSummary from '../../components/ExpensesSummary';
import {render, screen} from '@testing-library/react';
import {Provider} from 'react-redux';
import {expenses} from '../fixtures/expenses';
import appStore from '../../store/appStore';
import { addExpense } from '../../feature/expenseSlice';

//Mocks nanoid module because it's imported and used
jest.mock("nanoid", () => {
  return { nanoid: () => "1234" };
});

it('Should render with 1 expense added', () => {
    const store = appStore();
    const component = (
        <Provider store={store}>
            <ExpensesSummary expensesCount={1} expensesTotal={10}/>
        </Provider>
    );
    const {asFragment} = render(component);
    expect(asFragment()).toMatchSnapshot();
});

it('Should render with multiple expenses added', () => {
    const store = appStore();
    const component = (
        <Provider store={store}>
            <ExpensesSummary expensesCount={3} expensesTotal={50010} />
        </Provider>
    );
    const {asFragment} = render(component);
    expect(asFragment()).toMatchSnapshot();
});