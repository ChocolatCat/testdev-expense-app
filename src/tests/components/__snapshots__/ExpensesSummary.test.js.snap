// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render with 1 expense added 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Viewing 1 expense totalling $10.
    </h1>
  </div>
</DocumentFragment>
`;

exports[`Should render with multiple expenses added 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Viewing 3 expenses totalling $50,010.
    </h1>
  </div>
</DocumentFragment>
`;
